### PR TITLE
Precompute transpose of y in BicubicSplineInterpolation to speed up derivative wrt x2

### LINKED ISSUE #8997: Slow derivative wrt x2 in BicubicSplineInterpolation
### Description of the enhancement or error report
Sampling the spline in `BicubicSplineInterpolation` to get the derivative wrt the `x2` variable is much slower than getting the derivative wrt `x1` due to the construction of column vectors from the 2D arrays for each sampled point. This is destroying any scalability for me.

This can be fixed by constructing a transpose of the data once and saving it to memory.

### Rationale for the enhancement or information for reproducing the error
Evaluating derivatives wrt `x2` is too slow. Using some memory can speed this up considerably so that it is on par with sampling the derivative wrt `x1`.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
Uses some extra memory, but significantly speeds up sampling of derivative wrt `x2`

### PR DESCRIPTION
**Pull request by cpgr**
Closes #8997

### CODE REVIEW

File: framework/include/utils/BicubicSplineInterpolation.h [hunk 1]
```diff
@@ -67,6 +67,8 @@ class BicubicSplineInterpolation : public SplineInterpolationBase
   std::vector<Real> _x1;
   std::vector<Real> _x2;
   std::vector<std::vector<Real>> _y;
```
Comment (dschwen) [inline framework/include/utils/BicubicSplineInterpolation.h]: We have `DenseMatrix` for this purpose (it also has a get_transpose function)

File: framework/src/utils/BicubicSplineInterpolation.C [hunk 2]
```diff
@@ -114,22 +114,26 @@ BicubicSplineInterpolation::constructRowSplineSecondDerivativeTable()
 void
 BicubicSplineInterpolation::constructColumnSplineSecondDerivativeTable()
 {
+  auto m = _x1.size();
   auto n = _x2.size();
   _y2_columns.resize(n);
+  _y_trans.resize(n);
+
+  for (decltype(n) j = 0; j < n; ++j)
+    _y_trans[j].resize(m);
```
Comment (dschwen) [inline framework/src/utils/BicubicSplineInterpolation.C]: Yeah, all this manual resizing stuff would be unnecessary with `DenseMatrix`

File: framework/include/utils/BicubicSplineInterpolation.h (hunk 1 above)
Comment (dschwen) [inline framework/include/utils/BicubicSplineInterpolation.h]: Ugh nevermind, `spline` needs a vector anyways

**Comment by permcody** [comment]
Daniel, I'll assign this to you since you've already looked at it. Feel free to merge if it passes your review.

File: framework/src/utils/BicubicSplineInterpolation.C [hunk 3]
```diff
@@ -114,22 +114,26 @@ BicubicSplineInterpolation::constructRowSplineSecondDerivativeTable()
...
 
   // transpose the _y values so the columns can be easily iterated over
-  std::vector<std::vector<Real>> y_trans(_y[0].size(), std::vector<Real>(_y.size()));
```
Comment (dschwen) [inline framework/src/utils/BicubicSplineInterpolation.C]: :+1: for fewer temporary vector construction/destruction

**Review by dschwen — APPROVED** [review]

**Comment by cpgr** [comment]
Thanks @dschwen

**Pull request merged**